Adds the debugging

jmanero-r7 opened the pull request:
This PR encapsulates the work required to deploy a canary instance of turnstile.

Build/Deploy
- Adds a recipe to deploy from a GitHub HEAD
- Adds a recipe to deploy from a GitHub release artifact
- Adds a recipe to utilize a local copy of source, e.g. a synced VM folder

Logging and Audit
- Adds a configurable correlation identifier header
- Adds lots of debug logging
- Includes request identifiers in request-scoped logging
- Set default logging format to JSON

Misc
- Improves HTTPError classes' interfaces, allowing additional metadata fields to be passed to constructors
- Add a hacky curl-like CLI utility for testing the Turnstile interface
- LINT!

TravisCI
- Install build-essential and uuid-dev packages
- Make it build in an Ubuntu 14.04 environment for header/linking compatibility

jmanero-r7 committed:
Add alternate installation methods to support local debugging

* Add an enable flag to select service actions and gate notifications
* Add an attribute to select the installation method (default github)

jmanero-r7 committed:
Add a correlation identifier controller

* Use node-libuuid (bindings to uuid-dev) to generate IDs for
  un-identified requests

jmanero-r7 committed:
Add debug logging to controllers

* Add logic to the signature module to support both server and client
  request objects (`url` and `path` parameters, respectively)
* Ensure that the request method is uppercase

jmanero-r7 committed:
Log unhandled request errors to debug

jmanero-r7 committed:
Clean up detection/generation logic

* Add identifier tag to log output

jmanero-r7 committed:
Allow metadata hashes to be passed to Error types

jmanero-r7 committed:
Add request identifer to errors

jmanero-r7 committed:
Configure logging to generate flat JSON

jmanero-r7 committed:
Add package preconditions to travis config

jmanero-r7 committed:
Maybe install packages like this?

jmanero-r7 committed:
Annnd use ubuntu 14.04

jmanero-r7 committed:
Add support for installation from a non-released GitHub archive

jmanero-r7 committed:
Ensure that the turnstile directory exists

* Add logic to ensure that the archive is unpacked

jmanero-r7 committed:
Ensure that the target directory exists _before_ triggering the extract resource

jmanero-r7 committed:
Add a curl-like CLI utility for making signed requests to a turnstile endpoint

jmanero-r7 committed:
Ensure tha build-essential is installed for libuuid native build

jmanero-r7 committed:
Set sudo: required to use the trusty distribution

jmanero-r7 committed:
Update linting to use the Rapid7 Style Guide

Replace the airbnb/rapid7 configurations/plugins with the updated eslint-config-rapid7
rulesets and fix violations

jmanero-r7 committed:
Try removing eslint from root dependencies?

The CI is complaining that the eslint runtime can't load
eslint-plugin-import, which is depended upon in the
eslint-config-rapid7. The eslint-config-rapid7 package also
depends on a later version of eslint, which _may_ be resulting
in a peer-child dependency mismatch between the eslint and plugin
packages installed by eslint-config-rapid7 and that which we're
removing from the root `package.json`

jmanero-r7 committed:
Reverting 654f61d. That didn't work

jmanero-r7 committed:
Refactor the testing CLI into a less hacky script

that also happens to pass linting checks...

dgreene-r7 reviewed (changes requested):
21 commits across 20 files is way too big for a single PR. Can you split this into 4 separate ones that encompass each one of the topic areas you mention in the PR text?

jmanero-r7 committed:
Add a custom logging transport

to match Java/Logback formatting. Basically just
capitalizes the level and fills in some default configurations.

rvanburen-r7 commented on lib/log.js:
```diff
@@ -1,12 +1,61 @@
+   */
+  constructor(options) {
+    super(Object.assign({
+      colorize: false,
```
A dumb question, but why did you add Wonston.addColors but set colorize to false?

jmanero-r7 replied:
When the logger config is set to `json: false, colorize: true` from other configuration, the internal colorize method throws uncaught errors unless there's a matching mapping

rvanburen-r7 replied:
That makes sense, Thanks!

rvanburen-r7 commented on bin/turnt:
```diff
@@ -0,0 +1,254 @@
+    timestamp: false,
+    colorize: true
+  }
+  // log: {
```
Did you want to leave these commented out options here when they are specified in lib/config.js?

rvanburen-r7 commented on test/resource/http.js:
```diff
@@ -93,7 +92,7 @@ exports.server = function _listen(ready) {
     req.on('data', (chunk) => chunks.push(chunk));
     req.on('end', (chunk) => {
-      if (chunk) chunks.push(chunk);
+      if (chunk) { chunks.push(chunk); }
```
I don't know why this is causing your coverage to go down, but if you have a chance, you should get the coverage back up.

jmanero-r7 replied:
It's this addition... https://coveralls.io/builds/8416252/source?filename=lib%2Fsignature.js#L25 Somehow 3 lines account for 1.5 percent of the project 😒

rvanburen-r7 reviewed (approved):
While I agree with dgreene-r7, that this should have been broken up into 2 or 3 pull requests, I don't see anything wrong in the PR itself.

jmanero-r7 committed:
Clean up comment cruft from testing

fmitchell-r7 commented on cookbook/recipes/_install_github_ref.rb:
```diff
@@ -0,0 +1,42 @@
+           '--strip-components=1',
+           "--directory #{node['turnstile']['paths']['directory']}"].join(' ')
+
+  action ::File.exist?(
```
Can we leverage the `creates` attribute on [an execute resource](https://docs.chef.io/resource_execute.html), or are we using `action` because we want to trigger this when either the tarball is updated or the tarball failed to unpack?

jmanero-r7 replied:
The latter yes. AFAIK the guard clauses will squelch notifications as well as the configured action, so I went with this approach.

fmitchell-r7 commented on lib/errors.js:
```diff
@@ -73,6 +73,8 @@ class HTTPError extends Error {
     res.write(content);
     res.end();
+
+    Log.debug(err);
```
Is `err` always valid in this code path? If it comes in as `null` or `undefined` does the logic in lines 64 through 66 above still work?

jmanero-r7 replied:
Yes. `Error.handler` is only called internally by the controller: https://github.com/rapid7/turnstile/blob/master/lib/control/layer.js#L65-L70

fmitchell-r7 commented on lib/signature.js:
```diff
@@ -15,13 +15,31 @@ class Signature {
+    } else if (request.hasOwnProperty('path')) {
+      this.uri = request.path;
+    } else {
+      throw new Errors.RequestError('Missing URI field');
```
Is it worth capturing this `url` over `path` precedence in a test?

jmanero-r7 replied:
Tests added in 0e6a0db

fmitchell-r7 commented on lib/control/correlation.js:
```diff
@@ -0,0 +1,31 @@
+        identifier: req.identifier
+      });
+
+      return next();
```
Is it worth capturing this behavior of creating a correlation ID if one's not provided in a test?

jmanero-r7 replied:
Testes added in 01e6eeb

fmitchell-r7 commented on bin/turnt:
```diff
@@ -0,0 +1,247 @@
+    timestamp: false,
+    colorize: true
+  }
+});
```
Is it worth a few lines of documentation in the README about how to use this binary for testing?

jmanero-r7 replied:
I'm going reconfigure the docs directory in another PR

jmanero-r7 committed:
Add a set of tests for correlation-id middleware

jmanero-r7 committed:
Add test coverage for signature parameter detection

The Signature constructor allows the request URL to be passed
via the `url` or `path` parameters to support both ClientRequest
and IncomingMessage objects. This allows the same code-line to
be used to sign client-requests and verify incoming requests
on the server side.
